File: DesignPatternsExercises/java/FlyweightDesignPattern/FlyweightDesignPattern1/Solution/FlyweightDesignPattern1.java
import java.util.ArrayList;
import java.util.List;

public class FlyweightDesignPattern1 {
    public static void main(String[] args) {
        List<Tree> treeList = new ArrayList<>();

        for (int i = 0; i < 3; i++) {
            OakTree oakTree = (OakTree) TreeSelector.getTree("Oak Tree");
            oakTree.setRootDepth(i);
            treeList.add(oakTree);
        }

        for (int i = 0; i < 3; i++) {
            FloweringCherry floweringCherry = (FloweringCherry) TreeSelector.getTree("Flowering Cherry");
            treeList.add(floweringCherry);
        }

        for (Tree tree : treeList)
            System.out.println(tree);

        System.out.println("\nChanging property of a single Tree instance from the list at index 0");
        treeList.get(0).changeProperty();

        for (Tree tree : treeList)
            System.out.println(tree);
    }
}

File: DesignPatternsExercises/ruby/flyweight_design_pattern/flyweight_design_pattern1/exercise/flyweight_design_pattern1.rb
require_relative 'tree_selector'

tree_list = []

3.times do |depth|
  oak_tree = TreeSelector.tree('Oak Tree')
  oak_tree.root_depth = depth
  tree_list << oak_tree
end

3.times do
  flowering_cherry = TreeSelector.tree('Flowering Cherry')
  tree_list << flowering_cherry
end

tree_list.each { |tree| puts tree.to_string }

puts "\nChanging property of a single Tree instance from the list at index 0"
tree_list[0].change_property

tree_list.each { |tree| puts tree.to_string }
